### PR TITLE
parse events rather than txn.To.

### PR DESCRIPTION
**Pull request by puneet2019**
> The liquid staking contract is permissioned and only roles of user addresses are allowed
> Anchor batom contract is whitelisted to call the function

### CODE REVIEW

**Review by ajansari95 — APPROVED** [review]

File: ethereum/block.go [hunk 1]
```diff
@@ -124,3 +176,42 @@ func produceToKafka(kafkaProducer *sarama.SyncProducer) {
...
+}
+func ParseTokenWrapperWithdrawUTokensEvent(vLog types.Log) (tokenWrapper.TokenWrapperWithdrawUTokens, error) {
+	var event tokenWrapper.TokenWrapperWithdrawUTokens
+	err := contracts.TokenWrapper.GetABI().UnpackIntoInterface(&event, EventWithdrawUTokens, vLog.Data)
+	if err != nil {
+		return event, errors.New("failed to Unpack event StakeTokens")
+	}
+	event.Raw = vLog
+	event.AccountAddress = common.HexToAddress(vLog.Topics[1].Hex())
+	event.Tokens = vLog.Topics[2].Big()
```
Comment (aditim121) [inline ethereum/block.go:215]: what about event.finalTokens ?

File: ethereum/block.go [hunk 2]
```diff
@@ -124,3 +176,42 @@ func produceToKafka(kafkaProducer *sarama.SyncProducer) {
...
+	event.FinalTokens = vLog.Topics[3].Big()
+
+	return event, nil
+}
+func ParseLiquidStakingUnstakeTokensEvent(vLog types.Log) (liquidStaking.LiquidStakingUnstakeTokens, error) {
+	var event liquidStaking.LiquidStakingUnstakeTokens
+	err := contracts.LiquidStaking.GetABI().UnpackIntoInterface(&event, EventUnstakeTokens, vLog.Data)
+	if err != nil {
+		return event, errors.New("failed to Unpack event StakeTokens")
```
Comment (aditim121) [inline ethereum/block.go:199]: should be - failed to Unpack event UnstakeTokens

File: ethereum/block.go [hunk 3]
```diff
@@ -124,3 +176,42 @@ func produceToKafka(kafkaProducer *sarama.SyncProducer) {
...
+	event.AccountAddress = common.HexToAddress(vLog.Topics[1].Hex())
+	event.Tokens = vLog.Topics[2].Big()
+	event.FinalTokens = vLog.Topics[3].Big()
+	return event, nil
+}
+func ParseTokenWrapperWithdrawUTokensEvent(vLog types.Log) (tokenWrapper.TokenWrapperWithdrawUTokens, error) {
+	var event tokenWrapper.TokenWrapperWithdrawUTokens
+	err := contracts.TokenWrapper.GetABI().UnpackIntoInterface(&event, EventWithdrawUTokens, vLog.Data)
+	if err != nil {
+		return event, errors.New("failed to Unpack event StakeTokens")
```
Comment (aditim121) [inline ethereum/block.go:211]: should be - failed to Unpack event WithdrawUTokens

File: ethereum/block.go (hunk 2 above)
Comment (puneet2019) [reply ethereum/block.go:199]: fixed

File: ethereum/block.go (hunk 3 above)
Comment (puneet2019) [reply ethereum/block.go:211]: fixed

File: ethereum/block.go (hunk 1 above)
Comment (puneet2019) [reply ethereum/block.go:215]: That event is unindexed (for WithdrawUTokens) - so it is being parsed from UnpackInterface method. 

Manual parsing is required for events that are indexed.

Comment (aditim121) [reply ethereum/block.go:215]: got it

**Review by aditim121 — APPROVED** [review]

**Pull request merged**